File: common/src/main/kotlin/net/impleri/playerskills/data/conditions/ItemRestrictionConditionBuilder.kt
package net.impleri.playerskills.data.conditions

import com.google.gson.JsonObject
import net.impleri.playerskills.data.utils.RestrictionDataParser
import net.impleri.playerskills.restrictions.items.ItemConditions
import net.impleri.playerskills.restrictions.items.ItemRestriction
import net.minecraft.resources.ResourceLocation
import net.minecraft.world.entity.player.Player
import net.minecraft.world.item.Item

class ItemRestrictionConditionBuilder(name: ResourceLocation) :
  AbstractRestrictionConditionBuilder<Item, ItemRestriction>(name),
  ItemConditions<Player>,
  RestrictionDataParser {
  override var replacement: Item? = null
  override var producible: Boolean? = null
  override var consumable: Boolean? = null
  override var identifiable: Boolean? = null
  override var holdable: Boolean? = null
  override var wearable: Boolean? = null
  override var usable: Boolean? = true
  override var harmful: Boolean? = null

  override fun parseRestriction(
    jsonElement: JsonObject,
  ) {
    parseTarget(jsonElement, "item")
    producible = parseBoolean(jsonElement, "producible", producible)
    consumable = parseBoolean(jsonElement, "consumable", consumable)
    identifiable = parseBoolean(jsonElement, "identifiable", identifiable)
    holdable = parseBoolean(jsonElement, "holdable", holdable)
    wearable = parseBoolean(jsonElement, "wearable", wearable)
    usable = parseBoolean(jsonElement, "usable", usable)
    harmful = parseBoolean(jsonElement, "harmful", harmful)
  }

  override fun toggleEverything() {
    producible = true
    consumable = true
    identifiable = true
    holdable = true
    wearable = true
    usable = true
    harmful = true
  }

  override fun toggleNothing() {
    producible = false
    consumable = false
    identifiable = false
    holdable = false
    wearable = false
    usable = false
    harmful = false
  }
}

File: common/src/main/scala/net/impleri/playerskills/data/conditions/ItemRestrictionConditionBuilder.scala
package net.impleri.playerskills.data.conditions

import com.google.gson.JsonObject
import net.impleri.playerskills.api.skills.Skill
import net.impleri.playerskills.api.skills.SkillOps
import net.impleri.playerskills.api.skills.SkillType
import net.impleri.playerskills.api.skills.SkillTypeOps
import net.impleri.playerskills.restrictions.item.ItemConditions
import net.impleri.playerskills.server.api.Player
import net.impleri.playerskills.utils.PlayerSkillsLogger
import net.minecraft.resources.ResourceLocation

case class ItemRestrictionConditionBuilder(
  name: ResourceLocation,
  protected val skillOps: SkillOps = Skill(),
  protected val skillTypeOps: SkillTypeOps = SkillType(),
  protected val playerOps: Player = Player(),
  protected val logger: PlayerSkillsLogger = PlayerSkillsLogger.ITEMS,
) extends RestrictionConditionsBuilder with ItemConditions {
  override def parseRestriction(jsonElement: JsonObject): Unit = {
    parseTarget(jsonElement, "item")
    isIdentifiable = parseBoolean(jsonElement, "identifiable", isIdentifiable)
    isHoldable = parseBoolean(jsonElement, "holdable", isHoldable)
    isWearable = parseBoolean(jsonElement, "wearable", isWearable)
    isUsable = parseBoolean(jsonElement, "usable", isUsable)
    isHarmful = parseBoolean(jsonElement, "harmful", isHarmful)
  }

  override def toggleEverything(): Unit = {
    isIdentifiable = Option(true)
    isHoldable = Option(true)
    isWearable = Option(true)
    isUsable = Option(true)
    isHarmful = Option(true)
  }

  override def toggleNothing(): Unit = {
    isIdentifiable = Option(false)
    isHoldable = Option(false)
    isWearable = Option(false)
    isUsable = Option(false)
    isHarmful = Option(false)
  }
}
